ir/memory.cpp

```diff
@@ -887,14 +887,34 @@ expr Memory::alloc(const expr &size, unsigned align, BlockKind blockKind,
   (is_local ? local_blk_kind : non_local_blk_kind)
     .add(short_bid, expr::mkUInt(alloc_ty, 2));
 
-  return expr::mkIf(allocated, p(), Pointer::mkNullPointer(*this)());
+  return { p(), allocated };
 }
 
-void Memory::start_lifetime(const expr &ptr_local) {
+expr Memory::start_lifetime(const expr &ptr_local) {
   assert(!memory_unused());
   Pointer p(*this, ptr_local);
+
+  auto align = *local_blk_align(p.get_short_bid());
+  auto size_upperbound = p.block_size() + align.zextOrTrunc(bits_size_t) -
+                         expr::mkUInt(1, bits_size_t);
+  auto allocated = size_upperbound.ule(local_avail_space.zext(1));
+
+  if (observes_addresses()) {
+    // Disjointness of block's address range with other local blocks
+    auto disj = p.get_address().add_no_uoverflow(p.block_size()) &&
```

Comment:
I've changed alloc() such that this condition is always added regardless of alloc result. In summary, this can be removed from here.